examples/admission_controller.rs

```diff
@@ -0,0 +1,87 @@
+use std::convert::{Infallible, TryInto};
+
+use kube::api::{
+    admission::{AdmissionRequest, AdmissionResponse, AdmissionReview},
+    DynamicObject,
+};
+#[macro_use] extern crate log;
+use warp::{reply, Filter, Reply};
+
+#[tokio::main]
+async fn main() {
+    std::env::set_var("RUST_LOG", "debug,kube=debug");
+    tracing_subscriber::fmt::init();
+
+    let routes = warp::path("mutate")
+        .and(warp::body::json())
+        .and_then(mutate)
+        .with(warp::trace::request());
+
+
+    // For admission controllers running behind a service, you must generate a
+    // certificate for the service and encode the CA in the
+    // MutatingWebhookConfiguration, then terminate TLS here. For controllers
+    // running outside of the cluster, while still necessary, it's fine to
+    // handle TLS upstream. See:
+    // https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#tls-certificates
+
+    // warp::serve(warp::post().and(routes))
+    //     .tls()
+    //     .cert_path("admission-controller-tls.crt")
+    //     .key_path("admission-controller-tls.key")
+    //     .run(([0, 0, 0, 0], 8443))
+    //     .await;
+
+    warp::serve(warp::post().and(routes))
+        .run(([0, 0, 0, 0], 8080))
+        .await;
+}
+
+async fn mutate(body: AdmissionReview<DynamicObject>) -> Result<impl Reply, Infallible> {
+    let req: AdmissionRequest<DynamicObject> = match body.try_into() {
+        Ok(req) => req,
+        Err(err) => {
+            error!("invalid request: {}", err.to_string());
+            return Ok(reply::json(
+                &AdmissionResponse::invalid(err.to_string()).into_review(),
+            ));
+        }
+    };
+
+    let mut res = AdmissionResponse::from_request(&req);
+
+    if let Some(obj) = req.object {
```

Comment:
when is req.object not a Some? Is it perhaps reasonable to require it? Doesn't look like we can process the review without it.

Reply:
I'm unsure here, but the field is marked as optional in the protos https://github.com/kubernetes/api/blob/master/admission/v1/types.go#L97, so I was just being safe. I'm thinking only `old_object` is populated on `DELETE` requests

Reply:
Ah, ok. That makes sense.